### PR TITLE
[WIP]Chainerを使用したNeural networkによるMNIST分類器

### PR DESCRIPTION
**Pull request by YamaneSasuke**
numpyのNN実装をChainerで再現する．正解率の目標は98.2％

### CODE REVIEW

File: chainer_neural_network.py [hunk 1]
```diff
@@ -7,4 +7,123 @@
...
+            loss, accuracy = forward(model, x_batch, t_batch)
+            # 逆伝搬を計算
+            loss.backward()
+            optimizer.update()
+
+        time_end = time.time()
+        loss_train, accuracy_train = forward(model, X_train, T_train)
+        loss_valid, accuracy_valid = forward(model, X_valid, T_valid)
+
+        if accuracy_valid.data > accuracy_best:
```
Comment (ronekko) [inline chainer_neural_network.py]: 最善のモデルを確保しておく処理が欠けてます．
例えば `copy` モジュールの `deepcopy` 関数を使えば確保できます．
http://docs.python.jp/2/library/copy.html#copy.deepcopy

File: chainer_neural_network.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+        plt.plot(loss_trains)
+        plt.plot(loss_valids)
+        plt.title("loss")
+        plt.legend(["train", "valid"], loc="upper right")
+        plt.grid()
+        plt.show()
+
+    # テストデータの結果を表示
+    loss_test, accuracy_test = forward(model, x_test, t_test)
```
Comment (ronekko) [inline chainer_neural_network.py]: ここが `model` だと，単に最終エポック終了時点のモデルを使った評価になってしまうので， numpy版みたいに `best_model` を使うようになっている必要があります．

File: chainer_neural_network.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            # 勾配を初期化
+            optimizer.zero_grads()
+            # 順伝播を計算し、誤差と精度を取得
+            loss, accuracy = forward(model, x_batch, t_batch)
+            # 逆伝搬を計算
+            loss.backward()
+            optimizer.update()
+
+        time_end = time.time()
+        loss_train, accuracy_train = forward(model, X_train, T_train)
```
Comment (ronekko) [inline chainer_neural_network.py]: 現状のforward関数の実装では，ここでは `forward(..., train=False)` が必要になります（これがないと無意味な計算になる）．
もし誰かの実装を参考にしていて，さらにこのtrain引数の意味を理解せずに使ってるのだとしたらかなり危険です．
意味がわかってない道具（ここでは `F.dropout` ）を，何も考えずに使うべきではないです．なぜなら，うまくいったとしてもうまくいかなかったとしても，自分の中にあらかじめ予想を立ててないなら，練習として意味がないからです．
少なくとも， [公式レファレンス](http://docs.chainer.org/en/v1.4.1/reference/functions.html#chainer.functions.dropout)を読んで下さい．

File: chainer_neural_network.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        loss_valid, accuracy_valid = forward(model, X_valid, T_valid)
```
Comment (ronekko) [inline chainer_neural_network.py]: `forward(..., train=False)` が必要です．

File: chainer_neural_network.py (hunk 2 above)
Comment (ronekko) [inline chainer_neural_network.py]: `forward(..., train=False)` が必要です．

File: chainer_neural_network.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+# ニューラルネットワークの定義
+class MLP(Chain):
+    def __init__(self, num_features, dim_hidden, num_classes):
+        super(MLP, self).__init__(
+            l1=L.Linear(num_features, dim_hidden),
+            l2=L.Linear(dim_hidden, num_classes),
+        )
+
+    def loss_and_accuracy(self, X, T, train=False):
```
Comment (ronekko) [inline chainer_neural_network.py]: `train` 引数が不要です．
